File: benchmarks/jitbench.py
#!/usr/bin/env python3
"""
jitbench.py - Python equivalent of Limbo jitbench.b

Same 6 benchmarks with identical parameters for cross-language comparison.
CPython has no JIT - this is pure interpreter baseline.

Run:
    python3 jitbench.py
"""

import time

ITERATIONS = 10000000
SMALL_ITER = 1000000

# Python has arbitrary-precision integers. To match 64-bit signed wraparound
# behavior of C/Go/Java/Limbo, we mask to 64 bits where overflow can occur.
MASK64 = 0xFFFFFFFFFFFFFFFF

def to_signed64(v):
    """Convert masked 64-bit value to signed interpretation."""
    v = v & MASK64
    if v >= 0x8000000000000000:
        v -= 0x10000000000000000
    return v


def millisec():
    return int(time.monotonic() * 1000)


def warmup():
    s = 0
    for i in range(10000):
        s += i


def bench_arithmetic():
    a, b, c = 1, 2, 3
    mask = MASK64
    for _ in range(ITERATIONS):
        a = (a + b) & mask
        b = (b * 3) & mask
        c = (c - a) & mask
        a = (a ^ b) & mask
        b = b & 0xFFFF
        c = c | 0x1
        a = (a << 1) & mask
        b = b >> 1
        c = (c + (to_signed64(a) % 17)) & mask
    return to_signed64(a + b + c)


def bench_array():
    arr = list(range(1000))
    s = 0
    for _ in range(SMALL_ITER):
        for v in arr:
            s += v
    return s


def helper_add(a, b):
    return a + b


def bench_calls():
    s = 0
    for i in range(SMALL_ITER):
        s += helper_add(i, i + 1)
    return s


def fib(n):
    if n <= 1:
        return n
    return fib(n - 1) + fib(n - 2)


def bench_fib():
    s = 0
    for _ in range(100):
        s += fib(25)
    return s


def bench_sieve():
    SIZE = 100000
    count = 0
    for _ in range(10):
        sieve = [1] * SIZE
        sieve[0] = 0
        sieve[1] = 0
        i = 2
        while i * i < SIZE:
            if sieve[i]:
                for j in range(i * i, SIZE, i):
                    sieve[j] = 0
            i += 1
        count = sum(sieve)
    return count


def bench_nested():
    s = 0
    for i in range(500):
        for j in range(500):
            for k in range(200):
                s += i + j + k
    return s


def main():
    print("=== JIT Benchmark Suite (Python) ===")
    print(f"Iterations: {ITERATIONS} (arithmetic), {SMALL_ITER} (other)\n")

    warmup()

    t0 = millisec()

    print("1. Integer Arithmetic")
    t1 = millisec()
    result = bench_arithmetic()
    t2 = millisec()
    print(f"   Result: {result}, Time: {t2 - t1} ms\n")

    print("2. Loop with Array Access")
    t1 = millisec()
    result = bench_array()
    t2 = millisec()
    print(f"   Result: {result}, Time: {t2 - t1} ms\n")

    print("3. Function Calls")
    t1 = millisec()
    result = bench_calls()
    t2 = millisec()
    print(f"   Result: {result}, Time: {t2 - t1} ms\n")

    print("4. Fibonacci (recursive)")
    t1 = millisec()
    result = bench_fib()
    t2 = millisec()
    print(f"   Result: {result}, Time: {t2 - t1} ms\n")

    print("5. Sieve of Eratosthenes")
    t1 = millisec()
    result = bench_sieve()
    t2 = millisec()
    print(f"   Result: {result} primes, Time: {t2 - t1} ms\n")

    print("6. Nested Loops")
    t1 = millisec()
    result = bench_nested()
    t2 = millisec()
    print(f"   Result: {result}, Time: {t2 - t1} ms\n")

    tend = millisec()
    print(f"=== Total Time: {tend - t0} ms ===")


if __name__ == "__main__":
    main()

File: benchmarks/JITBench.java
/*
 * JITBench.java - Java equivalent of Limbo jitbench.b
 *
 * Same 6 benchmarks with identical parameters for cross-language comparison.
 * Uses long (64-bit) to match Limbo's 64-bit int.
 *
 * Compile and run:
 *   javac JITBench.java && java JITBench
 */

public class JITBench {

    static final int ITERATIONS = 10000000;
    static final int SMALL_ITER = 1000000;

    public static void main(String[] args) {
        System.out.println("=== JIT Benchmark Suite (Java) ===");
        System.out.printf("Iterations: %d (arithmetic), %d (other)%n%n", ITERATIONS, SMALL_ITER);

        warmup();

        // Warmup pass: run full suite once to let HotSpot compile hot loops
        benchArithmetic();
        benchArray();
        benchCalls();
        benchFib();
        benchSieve();
        benchNested();

        // Measurement pass
        long t0 = System.nanoTime();
        long result, t1, t2;

        System.out.println("1. Integer Arithmetic");
        t1 = System.nanoTime();
        result = benchArithmetic();
        t2 = System.nanoTime();
        System.out.printf("   Result: %d, Time: %d ms%n%n", result, (t2 - t1) / 1000000);

        System.out.println("2. Loop with Array Access");
        t1 = System.nanoTime();
        result = benchArray();
        t2 = System.nanoTime();
        System.out.printf("   Result: %d, Time: %d ms%n%n", result, (t2 - t1) / 1000000);

        System.out.println("3. Function Calls");
        t1 = System.nanoTime();
        result = benchCalls();
        t2 = System.nanoTime();
        System.out.printf("   Result: %d, Time: %d ms%n%n", result, (t2 - t1) / 1000000);

        System.out.println("4. Fibonacci (recursive)");
        t1 = System.nanoTime();
        result = benchFib();
        t2 = System.nanoTime();
        System.out.printf("   Result: %d, Time: %d ms%n%n", result, (t2 - t1) / 1000000);

        System.out.println("5. Sieve of Eratosthenes");
        t1 = System.nanoTime();
        result = benchSieve();
        t2 = System.nanoTime();
        System.out.printf("   Result: %d primes, Time: %d ms%n%n", result, (t2 - t1) / 1000000);

        System.out.println("6. Nested Loops");
        t1 = System.nanoTime();
        result = benchNested();
        t2 = System.nanoTime();
        System.out.printf("   Result: %d, Time: %d ms%n%n", result, (t2 - t1) / 1000000);

        long tend = System.nanoTime();
        System.out.printf("=== Total Time: %d ms ===%n", (tend - t0) / 1000000);
    }

    static void warmup() {
        long sum = 0;
        for (int i = 0; i < 10000; i++)
            sum += i;
        // Prevent DCE
        if (sum == Long.MIN_VALUE) System.out.print("");
    }

    static long benchArithmetic() {
        long a = 1, b = 2, c = 3;

        for (long i = 0; i < ITERATIONS; i++) {
            a = a + b;
            b = b * 3;
            c = c - a;
            a = a ^ b;
            b = b & 0xFFFF;
            c = c | 0x1;
            a = a << 1;
            b = b >> 1;
            c = c + (a % 17);
        }
        return a + b + c;
    }

    static long benchArray() {
        long[] arr = new long[1000];

        for (int i = 0; i < 1000; i++)
            arr[i] = i;

        long sum = 0;
        for (long j = 0; j < SMALL_ITER; j++) {
            for (int i = 0; i < 1000; i++)
                sum += arr[i];
        }
        return sum;
    }

    static long helperAdd(long a, long b) {
        return a + b;
    }

    static long benchCalls() {
        long sum = 0;
        for (long i = 0; i < SMALL_ITER; i++)
            sum += helperAdd(i, i + 1);
        return sum;
    }

    static long fib(long n) {
        if (n <= 1)
            return n;
        return fib(n - 1) + fib(n - 2);
    }

    static long benchFib() {
        long sum = 0;
        for (int i = 0; i < 100; i++)
            sum += fib(25);
        return sum;
    }

    static long benchSieve() {
        final int SIZE = 100000;
        long[] sieve = new long[SIZE];
        long count = 0;

        for (int iter = 0; iter < 10; iter++) {
            for (int i = 0; i < SIZE; i++)
                sieve[i] = 1;

            sieve[0] = 0;
            sieve[1] = 0;

            for (int i = 2; (long)i * i < SIZE; i++) {
                if (sieve[i] != 0) {
                    for (int j = i * i; j < SIZE; j += i)
                        sieve[j] = 0;
                }
            }

            count = 0;
            for (int i = 0; i < SIZE; i++)
                if (sieve[i] != 0)
                    count++;
        }
        return count;
    }

    static long benchNested() {
        long sum = 0;
        for (long i = 0; i < 500; i++)
            for (long j = 0; j < 500; j++)
                for (long k = 0; k < 200; k++)
                    sum += i + j + k;
        return sum;
    }
}
